go/libkb/pvl_interp.go

```diff
@@ -0,0 +1,825 @@
+
+		// Sanity check.
+		if int(state.PC) != i {
+			return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+				fmt.Sprintf("Execution failure, PC mismatch %v %v", state.PC, i))
+		}
+
+		newstate, perr := pvlStepInstruction(g, ins, state)
+		state = newstate
+		if perr != nil {
+			// Add additional info to INVALID_PVL errors.
+			if perr.GetProofStatus() == keybase1.ProofStatus_INVALID_PVL {
+				perr = NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					fmt.Sprintf("Invalid PVL (%v): %v", state.PC, perr.GetDesc()))
+			}
+			return perr
+		}
+		state.PC++
+	}
+
+	// Script executed successfully with no errors.
+	return nil
+}
+
+func pvlStepInstruction(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	for name, step := range PvlSteps {
+		if pvlJSONHasKey(ins, name) {
+			return step(g, ins, state)
+		}
+	}
+
+	return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+		"Unsupported PVL instruction %d", state.PC)
+}
+
+func pvlStepAssertRegexMatch(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	template, err := ins.AtKey(PvlAssertRegexMatch).GetString()
+	if err != nil {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not get template %v", ins)
+	}
+	re, perr := pvlInterpretRegex(template, state.Vars)
+	if perr != nil {
+		return state, perr
+	}
+	if !re.MatchString(state.ActiveString) {
+		g.GetLog().Debug("PVL regex did not match: %v %v", re, state.ActiveString)
+		return state, NewProofError(keybase1.ProofStatus_CONTENT_FAILURE,
+			"Regex did not match %v", re)
+	}
+
+	return state, nil
+}
+
+func pvlStepAssertFindBase64(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	target, err := ins.AtKey(PvlAssertFindBase64).GetString()
+	if err != nil {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not assert target %v", ins)
+	}
+	if target == "sig" {
+		if FindBase64Block(state.ActiveString, state.Vars.Sig, false) {
+			return state, nil
+		}
+		return state, NewProofError(keybase1.ProofStatus_TEXT_NOT_FOUND,
+			"Signature not found")
+	}
+	return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+		"Can only assert_find_base64 for sig")
+}
+
+func pvlStepWhitespaceNormalize(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	state.ActiveString = WhitespaceNormalize(state.ActiveString)
+	return state, nil
+}
+
+func pvlStepRegexCapture(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	template, err := ins.AtKey(PvlRegexCapture).GetString()
+	if err != nil {
+		return state, NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not get pattern %v", ins)
+	}
+	re, perr := pvlInterpretRegex(template, state.Vars)
+	if perr != nil {
+		return state, perr
+	}
+	match := re.FindStringSubmatch(state.ActiveString)
+	if len(match) < 2 {
```

Comment:
What's the significance of 2 here? This is the first match, followed by all of its subgroups, is that right? Are we asserting that it has at least one subgroup?